core/types/calls.cc

```diff
@@ -2451,6 +2451,140 @@ class Magic_selfNew : public IntrinsicMethod {
+            return;
+        }
+        auto fun = lit.asName(gs);
+
+        uint16_t numPosArgs = args.numPosArgs - 3;
+
+        InlinedVector<const TypeAndOrigins *, 2> sendArgStore;
+        InlinedVector<LocOffsets, 2> sendArgLocs;
+        for (int i = 3; i < args.args.size(); ++i) {
+            sendArgStore.emplace_back(args.args[i]);
+            sendArgLocs.emplace_back(args.locs.args[i]);
+        }
+        auto recvLoc = args.locs.args[0];
+        CallLocs sendLocs{args.locs.file, args.locs.call, recvLoc, args.locs.fun, sendArgLocs};
+
+        DispatchArgs innerArgs{
+            fun,
+            sendLocs,
+            numPosArgs,
+            sendArgStore,
+            selfTy.type,
+            selfTy,
+            selfTy.type,
+            args.block,
+            args.originForUninitialized,
+            args.isPrivateOk,
+            args.suppressErrors,
+        };
+        auto dispatched = selfTy.type.dispatchCall(gs, innerArgs);
+
+        auto multipleComponents = dispatched.secondary != nullptr;
+        if (multipleComponents) {
+            int unknownMethodOnNilClassErrors = 0;
+            for (auto it = &dispatched; it != nullptr; it = it->secondary.get()) {
+                for (auto &err : it->main.errors) {
+                    if (err->what == core::errors::Infer::UnknownMethod && it->main.receiver.isNilClass()) {
+                        unknownMethodOnNilClassErrors++;
+                    }
+                }
+            }
+
+            if (unknownMethodOnNilClassErrors == 1 && !core::Types::isSubType(gs, selfTy.type, selfTyAndAnd.type)) {
+                DispatchArgs newInnerArgs{
+                    fun,
+                    sendLocs,
+                    numPosArgs,
+                    sendArgStore,
+                    selfTyAndAnd.type,
+                    selfTyAndAnd,
+                    selfTyAndAnd.type,
+                    args.block,
+                    args.originForUninitialized,
+                    args.isPrivateOk,
+                    args.suppressErrors,
+                };
+                auto retried = selfTyAndAnd.type.dispatchCall(gs, newInnerArgs);
+
+                auto foundErrorOnRetry = false;
+                for (auto it = &retried; it != nullptr; it = it->secondary.get()) {
+                    foundErrorOnRetry |= !it->main.errors.empty();
+                }
+
+                if (!foundErrorOnRetry) {
+                    for (auto it = &dispatched; it != nullptr; it = it->secondary.get()) {
+                        for (auto &err : it->main.errors) {
+                            if (err->what == core::errors::Infer::UnknownMethod && it->main.receiver.isNilClass()) {
+                                if (auto newErr = gs.beginError(err->loc, core::errors::Infer::CallAfterAndAnd)) {
+                                    newErr.setHeader(
+                                        "Call to method `{}` after `{}` assumes result type doesn't change",
```

Comment:
I wonder if this phrasing will be a bit confusing, since "it's still nilable" doesn't exactly jump out here. Maybe something like "Receiver of call to method foo is still nilable even after &&"?